spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/ChangelogIterator.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.spark;
+
+import java.io.Serializable;
+import java.util.Iterator;
+import java.util.List;
+import org.apache.iceberg.ChangelogOperation;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.RowFactory;
+import org.apache.spark.sql.catalyst.expressions.GenericInternalRow;
+import org.apache.spark.sql.catalyst.expressions.GenericRowWithSchema;
+
+/**
+ * An iterator that transforms rows from changelog tables within a single Spark task.
+ *
+ * <p>It marks the carry-over rows to null to for filtering out later. Carry-over rows are unchanged
+ * rows in a snapshot but showed as delete-rows and insert-rows in a changelog table due to the
+ * copy-on-write(COW) mechanism. For example, there are row1 (id=1, data='a') and row2 (id=2,
```

Comment:
Thanks for comment.  Just wanted to check, why is it not possible for MOR?  Data file + position delete of same row?

Reply:
It may be precise to use `COW` here. Basically any metedata delete of a data file may have the carry-over rows. I will make the change.

Reply:
Thinking a bit more, MOR shouldn't have this carry-over rows. MOR will always delete a row which supposed to be deleted due to a delete or update operation. MOR won't delete other unrelated rows like COW does.

Reply:
Hm.. so if I do appendFile(dataFile) with a certain row, and then I do rowDelta() and add a position delete for that data file?  In the same snapshot (I think thats allowed?)  Is that considered a carry-over row?   Or is that taken care of some other way?